presto-orc/src/main/java/com/facebook/presto/orc/writer/KeyNode.java

```diff
@@ -0,0 +1,150 @@
+package com.facebook.presto.orc.writer;
+
+import static com.facebook.presto.orc.stream.LongOutputStream.createDataOutputStream;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.orc.DwrfDataEncryptor;
+import com.facebook.presto.orc.OrcEncoding;
+import com.facebook.presto.orc.OrcOutputBuffer;
+import com.facebook.presto.orc.checkpoint.BooleanStreamCheckpoint;
+import com.facebook.presto.orc.checkpoint.DoubleStreamCheckpoint;
+import com.facebook.presto.orc.checkpoint.LongStreamCheckpoint;
+import com.facebook.presto.orc.metadata.ColumnEncoding;
+import com.facebook.presto.orc.metadata.CompressedMetadataWriter;
+import com.facebook.presto.orc.metadata.CompressionParameters;
+import com.facebook.presto.orc.metadata.MetadataWriter;
+import com.facebook.presto.orc.metadata.RowGroupIndex;
+import com.facebook.presto.orc.metadata.Stream;
+import com.facebook.presto.orc.metadata.statistics.ColumnStatistics;
+import com.facebook.presto.orc.stream.BooleanOutputStream;
+import com.facebook.presto.orc.stream.LongOutputStream;
+import com.facebook.presto.orc.stream.PresentOutputStream;
+import com.facebook.presto.orc.stream.StreamDataOutput;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.slice.Slice;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class KeyNode implements ColumnWriter
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(KeyNode.class).instanceSize();
+    private final OrcOutputBuffer buffer;
+
+    private final int index;
+    private final ColumnWriter valueWriter;
+    private final BooleanOutputStream inMapStream;
+    private final CompressedMetadataWriter metadataWriter;
+    private long count;
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+
+    public KeyNode(int columnIndex, Type type, ColumnWriter valueWriter, CompressionParameters compressionParameters, Optional<DwrfDataEncryptor> dwrfEncryptor, MetadataWriter metadataWriter) {
+        index = columnIndex;
+        this.valueWriter = valueWriter;
+        inMapStream = new BooleanOutputStream(compressionParameters, dwrfEncryptor);
+        this.buffer = new OrcOutputBuffer(compressionParameters, dwrfEncryptor);
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, compressionParameters, dwrfEncryptor);
+    }
+
+    public void add(Block value) {
```

Comment:
this semantic is confusing to me. Block holds a list of value, but this is more of use like a single value. If single value is supposed to be added, we generally use the semantics add(Block value, int position)